### PR TITLE
Fix too lax application/json prefix-only matching

### PR DESCRIPTION
**Pull request by scop**
## What do these changes do?

Fix too lax application/json matching.

For example, `application/jsonfoo` or [`application/json-seq`](https://www.iana.org/assignments/media-types/application/json-seq) should not be treated as valid for JSON.

## Related issue number

Fixes https://github.com/aio-libs/aiohttp/issues/5896

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names.
- [x] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` for example (588.bugfix)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by scop** [comment]
Note that this will conflict with #6181, will fix the other after one of these is merged.

**Comment by asvetlov** [comment]
#6181 is merged, please fix conflicts

**Comment by scop** [comment]
Done/rebased.

File: tests/test_helpers.py
Comment (webknjaz) [inline tests/test_helpers.py]: Nit: instead of copy-pasting, I'd prefer having the same test be consolidated with `@pytest.mark.parametrize()`

File: tests/test_helpers.py
Comment (webknjaz) [inline tests/test_helpers.py]:
```suggestion
@pytest.mark.paratetrize(
    ("expected_ct", "response_ct", "should_match"),
    (
        ("application/json", "application/json; charset=UTF-8", True),
        ("application/json", "application/json-seq", False),
    ),
    ids=("with-charset", "non-plus-suffix-should-not-match"),
)
def test_is_expected_content_type_json(expected_ct, response_ct, should_match):
    assert is_expected_content_type(
        response_content_type=response_ct, expected_content_type=expected_ct
    ) is should_match
```

File: CHANGES/6180.bugfix
Comment (webknjaz) [inline CHANGES/6180.bugfix]:
```suggestion
Fixed matching the JSON media type to only accept ``application/json`` when it's plus-delimited -- :user:`scop`.
```

Comment (scop) [reply CHANGES/6180.bugfix]: I don't think this is what this change actually does -- there's nothing plus-delimited here. The only thing we do to the regex is to add the `(?:\s*;.*)?$` in the end (plus the resulting `black` formatting change).

Perhaps "Fixed matching the JSON media type to not accept arbitrary characters after ``application/json``."

Comment (webknjaz) [reply CHANGES/6180.bugfix]: Okay, that would do.

Comment (scop) [reply CHANGES/6180.bugfix]: Added that plus "or the ``+json`` media type suffix" as it applies to those types as well.

**Comment by scop** [comment]
Done.

File: aiohttp/helpers.py
Comment (webknjaz) [inline aiohttp/helpers.py:429]: Per PEP 257, there must always be an empty line between the docstring title and the description part.
```suggestion

    Both arguments should be given without parameters.
```

File: aiohttp/helpers.py
Comment (webknjaz) [inline aiohttp/helpers.py:127]: Maybe mark the start as well?
```suggestion
json_re = re.compile(r"^(?:application/|[\w.-]+/[\w.+-]+?\+)json$", re.IGNORECASE)
```

**Comment by webknjaz** [comment]
I don't have any additional comments beyond those few nitpicks so I'll let @asvetlov decide when to merge this.

File: aiohttp/helpers.py
Comment (scop) [reply aiohttp/helpers.py:127]: This regex is used with `.match` and is not a `re.MULTILINE` one, so the start marker does not make a difference. The majority regexps in aiohttp that are used in this kind of setup don't have it either. IMO adding it would just be a potential source of confusion.

(And IMO the whole `re.match` method shouldn't even exist, _it_ is just a source of confusion and `re.search` could be used instead of it everywhere. Maybe some day it annoys me enough that I'll start the journey trying to get it deprecated and eventually removed from Python. </rant>)

File: aiohttp/helpers.py
Comment (scop) [reply aiohttp/helpers.py:429]: BTW if PEP 257 is desired here, I suggest adding flake8-docstrings to pre-commit. I can submit a PR for that, let me know.

Comment (asvetlov) [reply aiohttp/helpers.py:429]: Maybe it is a good idea.
To the node: I'm not volunteered to fix all broken docstrings if the check is applied :)

File: aiohttp/helpers.py
Comment (asvetlov) [reply aiohttp/helpers.py:127]: Good to me

**Review by asvetlov — APPROVED** [review]

**Comment by asvetlov** [comment]
LGTM

**Pull request merged**
